### PR TITLE
Begin ECS Tutorial

### PR DESCRIPTION
**Pull request by sunwukonga**
Move chapters down to make room for `ecs tutorial`
Add mechanism for putting listing no. on code
Add mechanism for putting filename above code
Add link anchors to headings so they can be referenced
Add bevy tutorial SVG
Rough out the introduction to the tutorial
Fix wrong name for `add_system()` reference
  - there's no `add_system` on the IntoQuerySystem trait

### CODE REVIEW

**Comment by sunwukonga** [comment]
WIP: Not looking for this to be merged until the chapter is complete.

**Comment by sunwukonga** [comment]
This is what it looks like from top to bottom.

Screenshot_2020-08-30 Bevy - ecs tutorial

File: content/learn/book/contributing/docs/_index.md [hunk 1]
```diff
@@ -4,6 +4,7 @@ weight = 2
 sort_by = "weight"
 template = "book-section.html"
 page_template = "book-section.html"
+insert_anchor_links = "left"
```
Comment (cart) [inline content/learn/book/contributing/docs/_index.md:7]: left anchors make the heading alignment feel off. 

I'd prefer something closer to this:
image

```css
.bevy-link {
    font-size: 1.3rem;
    vertical-align: top;
    margin-left: 0.2rem;
}
```

```html
<a class="bevy-link" href="#{{ id }}" aria-label="Anchor link for: {{ id }}">🔗</a>
```

```
insert_anchor_links = "right"
```

File: content/learn/book/ecs/_index.md [hunk 2]
```diff
@@ -0,0 +1,114 @@
++++
+title = "ecs tutorial"
```
Comment (cart) [inline content/learn/book/ecs/_index.md:2]: Lets use this convention for titles: "ECS Tutorial"

File: content/learn/book/ecs/_index.md [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+  - with `commands:` {{rust_type(type="struct", crate="bevy_ecs", name="Commands", no_mod=true, plural=false)}}
+* {{rust_type(type="struct", crate="bevy_app", name="App", no_mod=true, plural=true)}} also have a mechanism for exiting gracefully
+
+## Create Project
+
+First [create](@/learn/book/getting-started/setup/_index.md#create-a-new-rust-executable-project) a new project called `bevy_ecs_tut` and [add](@/learn/book/getting-started/setup/_index.md#add-bevy-to-your-project-s-cargo-toml) <img src="/assets/bevy_logo_dark.svg" style="height: 1em;margin-bottom:-0.1em" alt="BEVY"/> as a dependency.
+
+If you don't know or remember the current version of Bevy, `cargo` can fetch it for you fast.
+```bash
+$ cargo search bevy
```
Comment (cart) [inline content/learn/book/ecs/_index.md:41]: We already have a section on setting up a bevy project with the latest version (which we linked to). I don't think we should duplicate content like this. Lets just link to the other section.

File: content/learn/book/ecs/_index.md [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+bevy = "0.1.3"                  # A refreshingly simple data-driven ...
+```
+{{caption(ref=3.1, desc="Truncated output of cargo search")}}
+
+## AppBuilder
```
Comment (cart) [inline content/learn/book/ecs/_index.md:46]: As a general (soft) rule, I think any code section should be preceded by some text explaining whats coming up. Dropping people right into code (especially a big block like this) can be a little overwhelming.

File: content/learn/book/ecs/_index.md [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        Total Players: 0
+        Winning Player: None
+```
+{{caption(ref=3.3, desc="First game output")}}
+
+The values of `GameState` come from the derived `Default` implementation.
+
+{{rust_type(type="struct" crate="bevy_app", name="App" method="build" no_struct=false)}} makes an {{rust_type(type="struct", crate="bevy_app", name="AppBuilder", no_mod=true, plural=false)}} for us, on which we can call inline methods to add **Systems**, **Components**, and configuration.
+
+{{rust_type(type="struct" crate="bevy_app", name="AppBuilder" method="init_resource" no_struct=false)}} inserts `GameState` into {{rust_type(type="struct", crate="bevy_app", name="App", no_mod=true, plural=false)}}::resources<{{rust_type(type="struct" crate="bevy_ecs", name="Resources", no_mod=false, plural=false)}}>, making `GameState` a **Component**.
```
Comment (cart) [inline content/learn/book/ecs/_index.md:92]: The formatting here feels a bit awkward / hard to parse:
image

I think we should make that a single "inline code block" and then add separate references after the fact.

File: content/learn/book/ecs/_index.md [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+If you don't know or remember the current version of Bevy, `cargo` can fetch it for you fast.
+```bash
+$ cargo search bevy
+bevy = "0.1.3"                  # A refreshingly simple data-driven ...
+```
+{{caption(ref=3.1, desc="Truncated output of cargo search")}}
+
+## AppBuilder
+
+{{fileref(name="src/main.rs")}}
```
Comment (cart) [inline content/learn/book/ecs/_index.md:48]: This is a really smart addition! Currently its a bit of a space hog though. Ideally it would overlay on top of the code block, but we sadly don't have direct control over code block elements. Instead I think we should do some margin hacking:

image

**Review by cart — CHANGES_REQUESTED** [review]
This is some great work! I really appreciate the time you put into this.

I want to preface this review with a disclaimer that I take docs very seriously and I will be pretty picky when it comes to presentation and content. Please don't interpret these comments being hostile or a "negative review". I will be employing an opinionated "heavy handed" approach for all Bevy Book PRs.

File: content/learn/book/ecs/_index.md [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  - Startup systems run in no particular order within their **stage** due to parallel execution
+  - Normal systems run in no particular order within their **stage** due to parallel execution
+* Stages exist as an ordered collection
+* For finer control of the order of execution within a scheduler
+  - You can create as many NEW stages in the scheduler as you like
+  - New stages are added immediately *before* or *after* an existing stage (system defined or custom)
+* You can emit and consume custom events
+* You can add new **Components** and **Systems** to a running {{rust_type(type="struct", crate="bevy_app", name="App", no_mod=true, plural=false)}}, i.e. after `.run()`
+  - with `commands:` {{rust_type(type="struct", crate="bevy_ecs", name="Commands", no_mod=true, plural=false)}}
+* {{rust_type(type="struct", crate="bevy_app", name="App", no_mod=true, plural=true)}} also have a mechanism for exiting gracefully
```
[the remark this answers is not in the document]
Comment (sunwukonga) [reply content/learn/book/ecs/_index.md:33]: Yes, I made the intro deliberately dense because I'm using it as a roadmap to make sure I touch on all of those points. I'll definitely make it more incoming-user friendly when I've included these points throughout the tutorial. Maybe I'll do a chapter review section at the end that looks more like this.

I actually created a little MUD battle simulation (Player vs rats) to use as the basis for this tutorial, but it very quickly got out of hand. The things that make it interesting are also the things that make it too long.

File: content/learn/book/ecs/_index.md [hunk 8]
```diff
@@ -0,0 +1,114 @@
++++
+title = "ecs tutorial"
+weight = 3
+sort_by = "weight"
+template = "book-section.html"
+page_template = "book-section.html"
+insert_anchor_links = "left"
++++
+
+<div style="text-align:right;"><img src="/assets/bevy_tutor_icon.svg" style="height: 4em;" alt="BEVY"/></div>
```
[the remark this answers is not in the document]
Comment (sunwukonga) [reply content/learn/book/ecs/_index.md:10]: Yeah, I went off the logo myself. In fact, I've been going back and forth on whether just to call it a tutorial at all. I just want it to be code and example heavy, it doesn't have to be called a tutorial.

File: content/learn/book/contributing/docs/_index.md (hunk 1 above)
Comment (sunwukonga) [reply content/learn/book/contributing/docs/_index.md:7]: I started off on the right :) I didn't like the way it looked, so I pushed it to the left for a little more prominence. Putting it back on the right is fine.

I like this style of link icon, but there's no implementation in Unicode, so if you want SVG it's the perfect opportunity for me. Which shape would you prefer? I could do both and see how it pans out at font size.

internal_link

File: content/learn/book/ecs/_index.md (hunk 3 above)
Comment (sunwukonga) [reply content/learn/book/ecs/_index.md:41]: Do you mind if I drop the `cargo search bevy` tip into `getting-started`?

File: content/learn/book/ecs/_index.md (hunk 4 above)
Comment (sunwukonga) [reply content/learn/book/ecs/_index.md:46]: It's a good idea. I should shift some of the context given below up above the code.

My experience as a reader of the Rust book at https://doc.rust-lang.org/book, which I otherwise enjoyed, was that it was a bit wordy. I tended to read the code first, and only if I didn't understand what they were trying to say there would I read the preamble.

I'd love a mechanism for finding a balance between the two. Perhaps a very concise breakdown of a paragraph, short-and-sweet, followed by a longer explanation. If I could highlight the breakdown to draw the eye, that would be perfect. I'll experiment with that and see what you think.

File: content/learn/book/contributing/docs/_index.md (hunk 1 above)
Comment (cart) [reply content/learn/book/contributing/docs/_index.md:7]: I wish headings could just automatically be links, instead of requiring a link icon. If we can make that work, that would be my preference.

The link icons above are definitely a common design, but they're a bit noisy for my taste. I like the simplicity of the "infinity" style logo. Or maybe even just a `#` character ... thats minimal and matches the anchor link style. The `#` wouldn't even require an SVG.

In general I want to reduce "noise" as much as possible and use a minimal "text first" / "content first" approach.

File: content/learn/book/ecs/_index.md (hunk 3 above)
Comment (cart) [reply content/learn/book/ecs/_index.md:41]: Yup thats fine. Lets add it after the "this is the current Bevy version" section.

"You can also look up the current version using the following cargo command:"

File: content/learn/book/ecs/_index.md (hunk 6 above)
Comment (sunwukonga) [reply content/learn/book/ecs/_index.md:48]: It wasn't really necessary for this particular chapter, because I don't intend to break out into separate files. But in the future there's definitely be a chapter on structuring the files of a game, and it'd annoying not having the immediate context of which file a code block is refering to.

**Comment by sunwukonga** [comment]
I also put a CSS `scroll-margin-top: 35em;` on the caption so that following a link to it will scroll the caption approximately to the bottom of the page. Mileage will vary though and it's a bit hacky.

File: content/learn/book/contributing/docs/_index.md (hunk 1 above)
Comment (sunwukonga) [reply content/learn/book/contributing/docs/_index.md:7]: I replaced with a custom link svg, that looks like the unicode one. This means it can now take the $default_color. The downside is that the svg is not dynamically scalable within Zola. However, I used three sizes that correspond to h2 a svg {}, h3 a svg {}, and h4 a svg {} repectively. I hide the sizes that shouldn't display in each case (`templates/anchor-link.html`) Short story, the symbol varies in size with respect to each heading type.

**Comment by alice-i-cecile** [comment]
Check out #154, which renders this PR obsolete by greatly expanding and revamping the content of the book.

**Comment by cart** [comment]
I'd also like to apologize to @sunwukonga for some truly terrible levels of communication / responsiveness. You did a lot of great work here and I didn't keep you in the loop on where my thoughts were. My excuse is that this pr was created during the extremely hectic initial bevy release. I was completely overwhelmed by the influx of people / questions / prs / social media and I couldn't give Bevy Book work the priority it needed. I was also very protective of Bevy Book content at the time and wasn't quite ready to let other people change the tone and structure of the book. I should have communicated that to you and I'm really sorry for disrespecting your good intentions and quality work.
